components/builder-api-client/src/builder.rs

```diff
@@ -186,7 +189,11 @@ impl BuilderAPIClient {
         fs::create_dir_all(&dst_path)?;
         let file_name = response::get_header(&resp, X_FILENAME)?;
         let dst_file_path = dst_path.join(file_name);
-        let w = AtomicWriter::new(&dst_file_path)?;
+        let w = {
+            let mut w = AtomicWriter::new(&dst_file_path)?;
+            w.with_permissions(permissions);
+            w
+        };
```

Comment:
This could be slightly shortned to:
```rust
let mut w = AtomicWriter::new(&dst_file_path)?;
w.with_permissions(permissions);
let w = w;
```

Should `with_permissions` be a constructor? Which avoids the need for `w` ever being mutable.

Reply:
I kinda want `AtomicWriter` to be created with a Builder pattern, but that's a whole bunch of extra code to write.

A constructor might be a nice compromise. I'll play with it.

Reply:
I agree that a builder pattern might be nice.  I held off just because it felt silly with a single method, but if we ever get to 2 methods, then I think moving it to the builder pattern would be nice.

Reply:
Worked out pretty nicely: https://github.com/habitat-sh/habitat/pull/7545/commits/a63b8a0577fcc27dea88b67befe917c51dfd790c